flink-ml-lib/src/main/java/org/apache/flink/ml/common/datastream/AllReduceImpl.java

```diff
@@ -0,0 +1,287 @@
+
+        // taskId, chunkId, totalElements, partitionedArray
+        DataStream<Tuple4<Integer, Integer, Integer, double[]>> allReduceSum =
+                allReduceSend
+                        .partitionCustom(
+                                (chunkId, numPartitions) -> chunkId % numPartitions, x -> x.f0)
+                        .transform(
+                                "all-reduce-sum",
+                                new TupleTypeInfo<>(
+                                        BasicTypeInfo.INT_TYPE_INFO,
+                                        BasicTypeInfo.INT_TYPE_INFO,
+                                        BasicTypeInfo.INT_TYPE_INFO,
+                                        PrimitiveArrayTypeInfo.DOUBLE_PRIMITIVE_ARRAY_TYPE_INFO),
+                                new AllReduceSum())
+                        .name("all-reduce-sum");
+
+        return allReduceSum
+                .partitionCustom((taskIdx, numPartitions) -> taskIdx % numPartitions, x -> x.f0)
+                .transform(
+                        "all-reduce-recv", TypeInformation.of(double[].class), new AllReduceRecv())
+                .name("all-reduce-recv");
+    }
+
+    /**
+     * Splits each double array into multiple chunks and send each chunk to the corresponding
+     * partition.
+     */
+    private static class AllReduceSend
+            extends RichFlatMapFunction<double[], Tuple3<Integer, Integer, double[]>> {
+        boolean hasReceivedOneRecord = false;
+        double[] transBuf = new double[TRANSFER_BUFFER_SIZE];
+
+        @Override
+        public void flatMap(
+                double[] receivedElement, Collector<Tuple3<Integer, Integer, double[]>> out) {
+            if (hasReceivedOneRecord) {
+                throw new RuntimeException("The input cannot contain more than one double array.");
+            }
+            hasReceivedOneRecord = true;
+            int numTasks = getRuntimeContext().getNumberOfParallelSubtasks();
+
+            for (int taskId = 0; taskId < numTasks; taskId++) {
+                int startChunkId = getStartChunkId(taskId, numTasks, receivedElement.length);
+                int numChunksToHandle =
+                        getNumChunksByTaskId(taskId, numTasks, receivedElement.length);
+                for (int chunkId = startChunkId;
+                        chunkId < numChunksToHandle + startChunkId;
+                        chunkId++) {
+                    int bufStart = chunkId * TRANSFER_BUFFER_SIZE;
+                    System.arraycopy(
+                            receivedElement,
+                            bufStart,
+                            transBuf,
+                            0,
+                            getLengthOfChunk(chunkId, receivedElement.length));
+                    out.collect(Tuple3.of(chunkId, receivedElement.length, transBuf));
+                }
+            }
+        }
+    }
+
+    /** Aggregates partitioned array from other workers. */
```

Comment:
Can we also explain the output of this operator here? For example, `Aggregates partitioned array from other workers and broadcast the aggregated array to every worker`.